ipnsfs/dir.go

```diff
@@ -0,0 +1,347 @@
+package ipnsfs
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	dag "github.com/jbenet/go-ipfs/merkledag"
+	ft "github.com/jbenet/go-ipfs/unixfs"
+	ufspb "github.com/jbenet/go-ipfs/unixfs/pb"
+)
+
+var ErrNotYetImplemented = errors.New("not yet implemented")
+var ErrInvalidChild = errors.New("invalid child node")
+
+type Directory struct {
+	fs        *Filesystem
+	parent    childCloser
+	childDirs map[string]*Directory
```

Comment:
I see `lock` being used [here](https://github.com/jbenet/go-ipfs/pull/923/files#diff-85be37b82b3ffbe8d91c875595d79e07R239) to protected these maps as well. See my comment below.